### PR TITLE
feat: Add conversation history persistence to LLMHistoryManager

### PR DESCRIPTION
**Pull request by aeemayo**
## Problem Statement
- The OM1 agent loses all conversation history when restarted, preventing continuity in interactions. Users need the agent to remember previous conversations and context across restarts.

## Solution
- I implemented conversation history persistence in "LLMHistoryManager "with the following changes:

## Key Changes:
- I Added  save_history()  method to write conversation history to JSONL format
- I Added load_history() method to restore history on initialization
- History auto-saves after each LLM response
- History auto-loads on startup if previous file exists
- Files stored at config/memory/history_{config_name}.jsonl
- Graceful handling of edge cases (empty files, missing files, no config name)

## Testing
- I Added 5 new unit tests for persistence functionality
- All 15 tests pass (pytest tests/providers/test_llm_history_manager.py)

### CODE REVIEW

File: src/capabilities/__init__.py [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+    global _global_registry
+    if _global_registry is None:
+        _global_registry = CapabilityRegistry()
+    return _global_registry
+
+
+def reset_capability_registry() -> None:
+    """Reset the global capability registry (mainly for testing)."""
+    global _global_registry
+    _global_registry = None
```
[the remark this answers is not in the document]
Comment (aeemars) [reply src/capabilities/__init__.py:364]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2701672593)

File: src/providers/llm_history_manager.py [hunk 2]
```diff
@@ -88,6 +91,73 @@ def __init__(
...
+                    data = json.loads(line)
+                    self.history.append(
+                        ChatMessage(role=data["role"], content=data["content"])
+                    )
+            if self.history:
+                logging.info(
+                    f"Loaded {len(self.history)} messages from history: {path}"
+                )
+        except Exception as e:
+            logging.error(f"Error loading history from {path}: {e}")
```
[the remark this answers is not in the document]
Comment (aeemars) [reply src/providers/llm_history_manager.py:198]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2701672644)

File: tests/capabilities/__init__.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+    def test_to_dict(self, populated_registry):
+        """Test converting to dictionary."""
+        summary = CapabilitySummary(populated_registry)
+        data = summary.to_dict()
+
+        assert "available_actions" in data
+        assert "available_sensors" in data
+        assert "constraints" in data
+        assert "components" in data
+        assert len(data["components"]) == 3
```
[the remark this answers is not in the document]
Comment (aeemars) [reply tests/capabilities/__init__.py:319]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2701672772)

File: src/fuser/__init__.py [hunk 4]
```diff
@@ -79,6 +79,11 @@ def fuse(self, inputs: list[Sensor], finished_promises: list[T.Any]) -> str:
         if self.config.system_prompt_examples:
             system_prompt += "\n\nEXAMPLES:\n" + self.config.system_prompt_examples
 
+        # Inject runtime capabilities into agent context
+        capabilities_summary = self.config.get_capabilities_summary()
+        if capabilities_summary:
+            system_prompt += "\n\n" + capabilities_summary
```
[the remark this answers is not in the document]
Comment (aeemars) [reply src/fuser/__init__.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2701672674)

[the remark this answers is not in the document]
Comment (aeemars) [reply src/fuser/__init__.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2701672674)

File: src/runtime/single_mode/config.py [hunk 5]
```diff
@@ -274,9 +297,88 @@ def load_config(
...
     parsed_config["cortex_llm"] = cortex_llm
 
+    # Build capability registry from loaded components
+    capability_registry = _build_capability_registry(
+        parsed_config["agent_actions"],
+        parsed_config["agent_inputs"],
+    )
+    parsed_config["capability_registry"] = capability_registry
+
     return RuntimeConfig(**parsed_config)
```
[the remark this answers is not in the document]
Comment (aeemars) [reply src/runtime/single_mode/config.py:306]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2701672682)

File: tests/providers/test_llm_history_manager.py [hunk 6]
```diff
@@ -358,3 +358,144 @@ async def process(self, prompt: str, messages: list) -> MagicMock:
...
+    config.history_length = 5
+    config.agent_name = "TestBot"
+    config.config_name = "test_config"
+
+    client = AsyncMock()
+    history_manager = LLMHistoryManager(config, client)
+
+    # Manually set the history path to use tmp_path
+    history_file = tmp_path / "history_test_config.jsonl"
+    history_manager._get_history_path = lambda: str(history_file)
```
[the remark this answers is not in the document]
Comment (aeemars) [reply tests/providers/test_llm_history_manager.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2705766652)

File: tests/providers/test_llm_history_manager.py [hunk 7]
```diff
@@ -358,3 +358,144 @@ async def process(self, prompt: str, messages: list) -> MagicMock:
...
+        f.write(json.dumps({"role": "user", "content": "Previous message"}) + "\n")
+        f.write(json.dumps({"role": "assistant", "content": "Previous response"}) + "\n")
+
+    client = AsyncMock()
+    history_manager = LLMHistoryManager(config, client)
+
+    # Override path to use tmp_path
+    history_manager._get_history_path = lambda: str(history_file)
+    history_manager.history = []  # Clear any loaded history
+    history_manager.load_history()
```
[the remark this answers is not in the document]
Comment (aeemars) [reply tests/providers/test_llm_history_manager.py]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2705766677)

File: src/providers/llm_history_manager.py [hunk 8]
```diff
@@ -88,6 +91,89 @@ def __init__(
...
+            Path to the history JSONL file, or None if no config_name is set.
+        """
+        if not self.config_name:
+            return None
+        # Sanitize config_name to avoid directory traversal or unsafe paths
+        safe_name = os.path.basename(self.config_name)
+        if os.path.sep in safe_name:
+            safe_name = safe_name.replace(os.path.sep, "_")
+        if os.path.altsep:
+            safe_name = safe_name.replace(os.path.altsep, "_")
```
[the remark this answers is not in the document]
Comment (aeemars) [reply src/providers/llm_history_manager.py:113]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2705766691)

File: src/providers/llm_history_manager.py [hunk 9]
```diff
@@ -88,6 +91,89 @@ def __init__(
...
+        if not path:
+            return
+
+        temp_path = f"{path}.tmp"
+        try:
+            os.makedirs(os.path.dirname(path), exist_ok=True)
+            with open(temp_path, "w", encoding="utf-8") as f:
+                for msg in self.history:
+                    f.write(
+                        json.dumps({"role": msg.role, "content": msg.content}) + "\n"
```
[the remark this answers is not in the document]
Comment (aeemars) [reply src/providers/llm_history_manager.py:135]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2705766703)

File: src/providers/llm_history_manager.py [hunk 10]
```diff
@@ -88,6 +91,89 @@ def __init__(
...
+        path = self._get_history_path()
+        if not path or not os.path.exists(path):
+            return
+
+        try:
+            with open(path, "r", encoding="utf-8") as f:
+                for line in f:
+                    line = line.strip()
+                    if not line:
+                        continue
```
[the remark this answers is not in the document]
Comment (aeemars) [reply src/providers/llm_history_manager.py:165]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2705766711)

File: tests/capabilities/test_base_classes.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
+"""
+Tests for capability introspection in base classes.
+"""
+
+from typing import Optional
```
[the remark this answers is not in the document]
Comment (aeemars) [reply tests/capabilities/test_base_classes.py:5]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2705751195)

File: tests/capabilities/test_base_classes.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+from actions.base import ActionConfig, ActionConnector
```
[the remark this answers is not in the document]
Comment (aeemars) [reply tests/capabilities/test_base_classes.py:7]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2705751211)

File: src/llm/__init__.py [hunk 13]
```diff
@@ -51,6 +51,9 @@ class LLMConfig(BaseModel):
     history_length: T.Optional[int] = Field(
         default=0, description="Number of past interactions to keep in context"
     )
+    config_name: T.Optional[str] = Field(
+        default=None, description="Name of the config file for history persistence"
+    )
```
[the remark this answers is not in the document]
Comment (aeemars) [reply src/llm/__init__.py:56]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2705784450)

File: src/runtime/single_mode/config.py [hunk 14]
```diff
@@ -89,6 +99,19 @@ class RuntimeConfig:
...
+        Get a formatted summary of runtime capabilities for agent context.
+
+        Returns
+        -------
+        str
+            Formatted capability summary string, or empty string if no registry
+        """
+        if self.capability_registry is None:
+            return ""
+        return self.capability_registry.generate_summary().to_prompt_string()
```
[the remark this answers is not in the document]
Comment (aeemars) [reply src/runtime/single_mode/config.py:112]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2705784455)

File: src/providers/llm_history_manager.py [hunk 15]
```diff
@@ -88,6 +91,89 @@ def __init__(
...
+                    data = json.loads(line)
+                    self.history.append(
+                        ChatMessage(role=data["role"], content=data["content"])
+                    )
+            if self.history:
+                logging.info(
+                    f"Loaded {len(self.history)} messages from history: {path}"
+                )
+        except Exception as e:
+            logging.error(f"Error loading history from {path}: {e}")
```
[the remark this answers is not in the document]
Comment (aeemars) [reply src/providers/llm_history_manager.py:198]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2705793066)

File: src/runtime/single_mode/config.py [hunk 16]
```diff
@@ -9,6 +9,11 @@
 from actions.base import AgentAction
 from backgrounds import load_background
 from backgrounds.base import Background
+from capabilities import (
+    CapabilityDescriptor,
+    CapabilityRegistry,
+    ComponentType,
+)
 from inputs import load_input
```
[the remark this answers is not in the document]
Comment (aeemars) [reply src/runtime/single_mode/config.py:17]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2705793081)

File: tests/capabilities/test_base_classes.py [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+        config = SensorConfig()
+        sensor = CapableSensor(config)
+        caps = sensor.get_capabilities()
+
+        assert caps is not None
+        assert caps.component_name == "capable_sensor"
+        assert caps.component_type == ComponentType.SENSOR
+        assert "sense1" in caps.supported_features
+        assert caps.get_constraint("update_rate").value == 30
+        assert caps.get_constraint("update_rate").unit == "Hz"
```
[the remark this answers is not in the document]
Comment (aeemars) [reply tests/capabilities/test_base_classes.py:112]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2709334262)

File: src/actions/base.py [hunk 18]
```diff
@@ -92,6 +95,21 @@ async def connect(self, output_interface: OT) -> None:
...
+        Override this method to expose runtime capabilities, supported features,
+        and constraints. If not overridden, returns None indicating no
+        capability information is available.
+
+        Returns
+        -------
+        Optional[CapabilityDescriptor]
+            A descriptor of this connector's capabilities, or None if not implemented
+        """
+        return None
```
[the remark this answers is not in the document]
Comment (aeemars) [reply src/actions/base.py:157]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2709334283)

File: src/actions/move/connector/ros2.py [hunk 19]
```diff
@@ -12,6 +14,34 @@ class MoveUnitreeSDKConnector(ActionConnector[ActionConfig, MoveInput]):
...
+                Constraint(name="max_speed", value=1.2, unit="m/s"),
+                Constraint(name="max_rotation", value=45, unit="deg/s"),
+            ],
+            is_available=True,
+            description="Movement control for Unitree robot via ROS2",
+            metadata={
+                "connector_type": "ros2",
+                "robot_type": "unitree",
+            },
+        )
```
[the remark this answers is not in the document]
Comment (aeemars) [reply src/actions/move/connector/ros2.py:51]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/OpenMind/OM1/pull/1252#discussion_r2709334304)

**Comment by Wanbogang** [comment]
@aeemayo 
do you really know what you are doing? before I check it

**Comment by aeemars** [comment]
> @aeemayo do you really know what you are doing? before I check it

Sure, i do.
